### DISCUSSION TITLE
Automatic plugin injection

### DISCUSSION THREAD

**Discussion by danielroe** [nuxt/nuxt#18189] [Ideas]
## Background

We now have a number of automatic features:
* automatic component registration & type recognition (via `GlobalComponents`)
* automatic composables/helpers registration & type declarations
* automatic server-api typings
* automatic response handling for direct returns from server middleware

This means that the typing of injected properties from plugins (via `nuxtApp.provide`) is one of the few areas left where users still have to manually augment types of their application. See issues (for example, https://github.com/nuxt/framework/issues/1863), and [current docs](https://v3.nuxtjs.org/docs/directory-structure/plugins#typing-plugins).

## Proposal

Users can return an object from a plugin to inject these values into the context, with the keys of the object being the keys injected. 

A key benefit of this approach is that we can automatically generate the types for these injected properties (including `$router`, for example - see https://github.com/nuxt/framework/issues/1794).

### Example
`~/plugins/auth.ts`:
```js
export default defineNuxtPlugin(() => {
  return {
    auth: {
      login: () => 'success'
    }
  }
})
```

`index.vue`
```vue
<template>
  {{ $auth.login() }}
</template>

<script setup>
// automatically typed
const { $auth } = useNuxtApp()
</script>
```

**Comment by pi0** [comment]
I like both ideas of being able to declare plugin injections and an easier way for lazy and nonblocking Nuxt plugins. However, there are a few notes to consider:

- Nuxt plugins are called **before rendering is started**. This gives them a chance to register the necessary hooks. By initializing them on the first usage during rendering, we are basically **breaking the Lifecycle of a Nuxt plugin** and purposed lazy behavior **can be already implemented by an auto imported composable**. 
- While this saves time for module/plugin authors, it will add verbose code for consumers to call `await $service.load()` before being able to use it. Imagine usage of a nested API like `$analytics.page.sendEvent()`
- Lazy and nonblocking patterns can be already implemented using dynamic imports and better control of plugin author. We can improve this area by providing more utils for plugin authors (see example 1)
- In order to improve type-hinting by automatically generating types, we have alternative ways.
  - Module authors, when using addPlugin, can set `{ provides: [ '$service'] }`
  - We can use static analyzes to extract meta with `export const meta = { provides: '$service' }`
  - => Also note that inferring injected name from the file name alone is tricky and wrong. This assumes **nuxt plugins exactly provide one service** that is identical to their name! This doesn't apply to all use cases.
  - => (We can do type inference based on return object instead see below)

BTW I like the idea of returning named exports to automatically provide them and infer the types **without lazy behavior inferring from filename** 👍 

```js
// plugins/log.ts
export default defineNuxtPlugin((nuxtApp) => {
  return {
    // Provides $log (same as nuxtApp.provide('log', { ... }))
    log: {
     info: (..args) => console.info(...args)
    }
  }
})
```

---

Example 1: Example of lazy implementation with dynamic imports:

```js
export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.provide('$big', {
    sendEvent: async (name) => {
      const { sendEvent } = await import('big-library')
      return sendEvent(nuxtApp, name)
    }
  })
})
```


```vue
<script setup>
await useNuxtApp().$big.sendEvent('hello')
</script>
```

**Reply by danielroe** [reply]
Nice comments! Will amend RFC.

I think this is correct approach:
>  I like the idea of returning named exports to automatically provide them and infer the types without lazy behavior inferring from filename 👍

Static analysis or meta wouldn't let us make best advantage of TS compiler. And also returning object simplifies job of type inference and naming.

**Reply by pi0** [reply]
> Static analysis or meta wouldn't let us make best advantage of TS compiler. And also returning object simplifies job of type inference and naming.

Agreed! Returning is probably better way to make it working out of the box. Thare are edge cases similar to `server/api` inference (like when one uses `req.send` instead of retuning but we can rethink about that later.

**Reply by danielroe** [reply]
Yes. And certainly for more complex cases users can continue to declare their own injections 👍
